File: ocaml/src/vectorize/loop_unrolling.ml
open Llvm
open Llvm_analysis
open Llvm_scalar_opts
open Llvm_target
open Llvm_transforms
open Llvm_ipo

module CL = Cmdliner

type loop = {
  header: llbasicblock;
  latch: llbasicblock;
  blocks: llbasicblock list;
  exits: llbasicblock list;
}

type unroll_loop_options = {
  count: int;
  trip_count: int;
  trip_multiple: int;
  peel_count: int;
  allow_runtime: bool;
  allow_expensive_trip_count: bool;
  force: bool;
  unroll_remainder: bool;
  forget_all_scev: bool;
}

type loop_unroll_result =
  | Unmodified
  | PartiallyUnrolled
  | FullyUnrolled

type unrolled_value = {
  iteration: int;
  original: llvalue;
}

let unroll_runtime_epilog = ref false

let need_to_insert_phis_for_lcssa l blocks li =
  List.exists (fun bb ->
    if Loop.contains l bb then false
    else
      Llvm.fold_left_instrs (fun acc instr ->
        acc || List.exists (fun op ->
          match value_as_instruction op with
          | Some def_instr ->
            let def_loop = Li.get_loop_for (instr_parent def_instr) in
            Option.is_some def_loop && Loop.contains def_loop l
          | None -> false
        ) (use_begin instr)
      ) false bb
  ) blocks

let simplify_loop_after_unroll2 l simplify_ivs li se dt ac tti =
  if se && simplify_ivs then begin
    let dead_insts = ref [] in
    simplify_loop_ivs l se dt li tti dead_insts;
    List.iter (fun v ->
      match value_as_instruction v with
      | Some inst -> delete_instruction inst
      | None -> ()
    ) !dead_insts
  end;
  
  let dl = module_data_layout (global_parent (block_parent l.header)) in
  List.iter (fun bb ->
    Llvm.iter_instrs (fun inst ->
      match simplify_instruction inst dl None dt ac with
      | Some v when Li.replacement_preserves_lcssa_form inst v ->
        replace_all_uses_with inst v
      | _ ->
        if is_instruction_trivially_dead inst then
          delete_instruction inst
    ) bb
  ) l.blocks

let is_epilog_profitable l =
  let preheader = Loop.get_loop_preheader l in
  let header = l.header in
  List.exists (fun phi ->
    match incoming_value phi preheader with
    | Some v -> is_constant v
    | None -> false
  ) (phi_nodes header)

let rec unroll_loop_with_vmap l ulo li se dt ac tti preserve_lcssa unroll_to_orig_map remainder_loop =
  let open Llvm in
  
  if not (Loop.get_loop_preheader l) then
    Unmodified
  else if not (Loop.get_loop_latch l) then
    Unmodified
  else if not (Loop.is_safe_to_clone l) then
    Unmodified
  else if has_address_taken l.header then
    Unmodified
  else begin
    let ulo = if ulo.trip_count <> 0 && ulo.count > ulo.trip_count then
      { ulo with count = ulo.trip_count }
    else ulo in

    if ulo.trip_count = 0 && ulo.count < 2 && ulo.peel_count = 0 then
      Unmodified
    else begin
      let completely_unroll = ulo.count = ulo.trip_count in
      let runtime_trip_count = ulo.trip_count = 0 && ulo.count > 0 && ulo.allow_runtime in

      let peeled = ref false in
      if ulo.peel_count > 0 then begin
        peeled := peel_loop l ulo.peel_count li se dt ac preserve_lcssa;
        if !peeled then begin
          let exiting_block = l.latch in
          let new_trip_count = Se.get_small_constant_trip_count l exiting_block in
          let new_trip_multiple = Se.get_small_constant_trip_multiple l exiting_block in
          ulo = { ulo with trip_count = new_trip_count; trip_multiple = new_trip_multiple }
        end
      end;

      let orig_loop_blocks = l.blocks in
      let breakout_trip = 
        if ulo.trip_count <> 0 then ulo.trip_count mod ulo.count
        else ulo.trip_multiple in

      if se then
        if ulo.forget_all_scev then Se.forget_all_loops se
        else Se.forget_topmost_loop se l;

      let rec clone_iterations count acc last_value_map =
        if count = 0 then (acc, last_value_map)
        else
          let new_blocks, new_value_map = 
            List.fold_left (fun (blocks, map) bb ->
              let new_bb, new_map = clone_basic_block_with_map bb in
              (new_bb :: blocks, ValueMap.union new_map map)
            ) ([], last_value_map) orig_loop_blocks
          in
          clone_iterations (count - 1) (new_blocks :: acc) new_value_map
      in
      let unrolled_blocks, last_value_map = clone_iterations ulo.count [] ValueMap.empty in

      List.iter (fun phi ->
        if completely_unroll then
          replace_all_uses_with phi (incoming_value phi (Loop.get_loop_preheader l))
        else if ulo.count > 1 then begin
          let in_val = remove_incoming phi l.latch in
          let in_val = 
            match value_as_instruction in_val with
            | Some instr when Loop.contains l (instr_parent instr) ->
              ValueMap.find in_val last_value_map
            | _ -> in_val
          in
          add_incoming phi in_val (List.hd (List.rev unrolled_blocks))
        end
      ) (phi_nodes l.header);

      List.iteri (fun i latch ->
        let dest = if i = List.length unrolled_blocks - 1 then l.header else List.nth unrolled_blocks (i + 1) in
        let term = terminator latch in
        match value_as_branch term with
        | Some br when is_conditional br ->
          set_successor br 0 dest
        | _ ->
          let new_br = build_br dest (instr_parent term) in
          replace_all_uses_with term new_br;
          delete_instruction term
      ) (List.map (fun blocks -> List.hd (List.rev blocks)) unrolled_blocks);

      simplify_loop_after_unroll2 l (not completely_unroll && (ulo.count > 1 || !peeled)) li se dt ac tti;

      if preserve_lcssa && completely_unroll then
        List.iter (fun exit_bb ->
          List.iter (fun phi ->
            List.iter (fun block ->
              if not (Loop.contains l block) then
                remove_incoming phi block
            ) (incoming_blocks phi)
          ) (phi_nodes exit_bb)
        ) l.exits;

      if completely_unroll then Li.erase l;

      if completely_unroll then FullyUnrolled else PartiallyUnrolled
    end
  end

File: go/vectorize/loop_unrolling.go
package hyper

import (
	"flag"
	"fmt"
	"log"

	"github.com/llir/llvm"
	"github.com/llir/llvm/ir"
	"github.com/llir/llvm/ir/constant"
	"github.com/llir/llvm/ir/types"
	"github.com/llir/llvm/ir/value"
)

type Loop struct {
	Header *ir.Block
	Latch  *ir.Block
	Blocks []*ir.Block
	Exits  []*ir.Block
}

type UnrollLoopOptions struct {
	Count                    int
	TripCount                int
	TripMultiple             int
	PeelCount                int
	AllowRuntime             bool
	AllowExpensiveTripCount  bool
	Force                    bool
	UnrollRemainder          bool
	ForgetAllSCEV            bool
}

type LoopUnrollResult int

const (
	Unmodified LoopUnrollResult = iota
	PartiallyUnrolled
	FullyUnrolled
)

type UnrolledValue struct {
	Iteration int
	Original  value.Value
}

var unrollRuntimeEpilog bool

func needToInsertPhisForLCSSA(l *Loop, blocks []*ir.Block, li *LoopInfo) bool {
	for _, bb := range blocks {
		if li.GetLoopFor(bb) == l {
			continue
		}
		for _, inst := range bb.Insts {
			for _, op := range inst.Operands() {
				if defInst, ok := op.(*ir.Instruction); ok {
					defLoop := li.GetLoopFor(defInst.Parent)
					if defLoop != nil && defLoop.Contains(l) {
						return true
					}
				}
			}
		}
	}
	return false
}

func simplifyLoopAfterUnroll2(l *Loop, simplifyIVs bool, li *LoopInfo, se *ScalarEvolution, dt *DominatorTree, ac *AssumptionCache, tti *TargetTransformInfo) {
	if se != nil && simplifyIVs {
		deadInsts := make([]*ir.Instruction, 0)
		simplifyLoopIVs(l, se, dt, li, tti, &deadInsts)
		for _, inst := range deadInsts {
			inst.Parent.Insts.Remove(inst)
		}
	}

	dl := l.Header.Parent.Module.DataLayout
	for _, bb := range l.Blocks {
		for i := 0; i < len(bb.Insts); i++ {
			inst := bb.Insts[i]
			if v := simplifyInstruction(inst, dl, nil, dt, ac); v != nil {
				if li.ReplacementPreservesLCSSAForm(inst, v) {
					inst.ReplaceAllUsesWith(v)
				}
			}
			if isInstructionTriviallyDead(inst) {
				bb.Insts.Remove(inst)
				i--
			}
		}
	}
}

func isEpilogProfitable(l *Loop) bool {
	preheader := l.Header.Parent.Blocks[len(l.Header.Parent.Blocks)-1]
	for _, phi := range l.Header.Insts {
		if p, ok := phi.(*ir.InstPhi); ok {
			for _, inc := range p.Incs {
				if inc.Pred == preheader {
					if _, ok := inc.X.(*constant.Int); ok {
						return true
					}
				}
			}
		}
	}
	return false
}

func unrollLoopWithVMap(l *Loop, ulo UnrollLoopOptions, li *LoopInfo, se *ScalarEvolution, dt *DominatorTree, ac *AssumptionCache, tti *TargetTransformInfo, preserveLCSSA bool, unrollToOrigMap map[value.Value]UnrolledValue, remainderLoop **Loop) LoopUnrollResult {
	if l.GetLoopPreheader() == nil {
		return Unmodified
	}

	if l.GetLoopLatch() == nil {
		return Unmodified
	}

	if !l.IsSafeToClone() {
		return Unmodified
	}

	if l.Header.HasAddressTaken() {
		return Unmodified
	}

	if ulo.TripCount != 0 && ulo.Count > ulo.TripCount {
		ulo.Count = ulo.TripCount
	}

	if ulo.TripCount == 0 && ulo.Count < 2 && ulo.PeelCount == 0 {
		return Unmodified
	}

	completelyUnroll := ulo.Count == ulo.TripCount
	runtimeTripCount := ulo.TripCount == 0 && ulo.Count > 0 && ulo.AllowRuntime

	peeled := false
	if ulo.PeelCount > 0 {
		peeled = peelLoop(l, ulo.PeelCount, li, se, dt, ac, preserveLCSSA)
		if peeled {
			exitingBlock := l.GetLoopLatch()
			ulo.TripCount = se.GetSmallConstantTripCount(l, exitingBlock)
			ulo.TripMultiple = se.GetSmallConstantTripMultiple(l, exitingBlock)
		}
	}

	origLoopBlocks := l.Blocks

	if se != nil {
		if ulo.ForgetAllSCEV {
			se.ForgetAllLoops()
		} else {
			se.ForgetTopmostLoop(l)
		}
	}

	var breakoutTrip int
	if ulo.TripCount != 0 {
		breakoutTrip = ulo.TripCount % ulo.Count
		ulo.TripMultiple = 0
	} else {
		breakoutTrip = ulo.TripMultiple
		ulo.TripMultiple = gcd(ulo.Count, ulo.TripMultiple)
	}

	lastValueMap := make(map[value.Value]value.Value)
	var origPHINode []*ir.InstPhi
	for _, inst := range l.Header.Insts {
		if phi, ok := inst.(*ir.InstPhi); ok {
			origPHINode = append(origPHINode, phi)
		}
	}

	var headers, latches []*ir.Block
	headers = append(headers, l.Header)
	latches = append(latches, l.GetLoopLatch())

	unrolledLoopBlocks := l.Blocks

	loopsToSimplify := make(map[*Loop]struct{})
	for _, subLoop := range l.SubLoops {
		loopsToSimplify[subLoop] = struct{}{}
	}

	for it := 1; it != ulo.Count; it++ {
		newBlocks := make([]*ir.Block, 0)
		newLoops := make(map[*Loop]*Loop)
		newLoops[l] = l

		for _, bb := range origLoopBlocks {
			newBB := cloneBasicBlock(bb, fmt.Sprintf(".%d", it))
			l.Header.Parent.Blocks = append(l.Header.Parent.Blocks, newBB)

			unrollToOrigMap[newBB] = UnrolledValue{Iteration: it, Original: bb}

			oldLoop := addClonedBlockToLoopInfo(bb, newBB, li, newLoops)
			if oldLoop != nil {
				loopsToSimplify[newLoops[oldLoop]] = struct{}{}
			}

			if bb == l.Header {
				for _, origPHI := range origPHINode {
					newPHI := newBB.Insts[0].(*ir.InstPhi)
					inVal := newPHI.Incs[len(newPHI.Incs)-1].X
					if inValI, ok := inVal.(*ir.Instruction); ok {
						if it > 1 && l.Contains(inValI.Parent) {
							inVal = lastValueMap[inValI]
						}
					}
					lastValueMap[origPHI] = inVal
					newBB.Insts = newBB.Insts[1:]
				}
			}

			for i, inst := range bb.Insts {
				lastValueMap[inst] = newBB.Insts[i]
				unrollToOrigMap[newBB.Insts[i]] = UnrolledValue{Iteration: it, Original: inst}
			}

			for _, succ := range bb.Term.Succs() {
				if l.Contains(succ) {
					continue
				}
				for _, phi := range succ.Insts {
					if p, ok := phi.(*ir.InstPhi); ok {
						incoming := p.Incoming(bb)
						if newIncoming, exists := lastValueMap[incoming]; exists {
							incoming = newIncoming
						}
						p.Incs = append(p.Incs, &ir.Incoming{Pred: newBB, X: incoming})
					}
				}
			}

			if bb == l.Header {
				headers = append(headers, newBB)
			}
			if bb == l.GetLoopLatch() {
				latches = append(latches, newBB)
			}

			newBlocks = append(newBlocks, newBB)
			unrolledLoopBlocks = append(unrolledLoopBlocks, newBB)

			if dt != nil {
				if bb == l.Header {
					dt.AddNewBlock(newBB, latches[it-1])
				} else {
					bbDomNode := dt.GetNode(bb)
					bbIDom := bbDomNode.IDom()
					originalBBIDom := bbIDom.Block()
					dt.AddNewBlock(newBB, lastValueMap[originalBBIDom].(*ir.Block))
				}
			}
		}

		remapInstructions(newBlocks, lastValueMap)
		for _, newBlock := range newBlocks {
			for _, inst := range newBlock.Insts {
				if ii, ok := inst.(*ir.InstCall); ok {
					if ii.Callee == l.Header.Parent.Module.NamedFuncs["llvm.assume"] {
						ac.RegisterAssumption(ii)
					}
				}
			}
		}
	}

	for _, pn := range origPHINode {
		if completelyUnroll {
			pn.ReplaceAllUsesWith(pn.Incoming(l.GetLoopPreheader()))
			l.Header.Insts.Remove(pn)
		} else if ulo.Count > 1 {
			inVal := pn.RemoveIncoming(l.GetLoopLatch())
			if inValI, ok := inVal.(*ir.Instruction); ok {
				if l.Contains(inValI.Parent) {
					inVal = lastValueMap[inVal]
				}
			}
			pn.Incs = append(pn.Incs, &ir.Incoming{Pred: latches[len(latches)-1], X: inVal})
		}
	}

	for i, latch := range latches {
		dest := headers[(i+1)%len(headers)]
		term := latch.Term.(*ir.TermBr)
		if term.Cond != nil {
			term.SuccTrue = dest
		} else {
			newBr := ir.NewBr(dest)
			latch.Term.ReplaceAllUsesWith(newBr)
			latch.Term = newBr
		}
	}

	simplifyLoopAfterUnroll2(l, !completelyUnroll && (ulo.Count > 1 || peeled), li, se, dt, ac, tti)

	outerL := l.Parent
	if completelyUnroll {
		li.Erase(l)
	}

	if preserveLCSSA && outerL != nil && completelyUnroll {
		needToFixLCSSA := needToInsertPhisForLCSSA(outerL, unrolledLoopBlocks, li)
		if needToFixLCSSA {
			latchLoop := li.GetLoopFor(latches[len(latches)-1])
			fixLCSSALoop := outerL
			for fixLCSSALoop.Parent != latchLoop {
				fixLCSSALoop = fixLCSSALoop.Parent
			}
			formLCSSARecursively(fixLCSSALoop, dt, li, se)
		}
	}

	if dt != nil {
		if outerL != nil {
			simplifyLoop(outerL, dt, li, se, ac, nil, preserveLCSSA)
		} else {
			for loop := range loopsToSimplify {
				simplifyLoop(loop, dt, li, se, ac, nil, preserveLCSSA)
			}
		}
	}

	if completelyUnroll {
		return FullyUnrolled
	}
	return PartiallyUnrolled
}